### DISCUSSION TITLE
`StatefulRearchConsumer`?

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#45] [Q&A] [answered]
Seems that `RearchConsumer` is like a `StatelessWidget` compared to a `StatefulWidget`.
It hasn't lifecycle methods like `initState` or `dispose`.
I will start tonight to refactor my project Flutter side to use `rearch` but seems maybe a problem to not be able to have `StatefulRearchWidget`.

Maybe I'm wrong and it exists...

**Comment by GregoryConrad** [comment, marked as answer]
> Seems that RearchConsumer is like a StatelessWidget compared to a StatefulWidget.

Well it is actually most comparable to `StatefulWidget`, but just abstracting the OOP away from you.

> It hasn't lifecycle methods like initState or dispose.

That's right, they are abstracted away from you via side effects. You should use side effects when in a RearchConsumer.

`initState` and `dispose` are handled by `use.callonce` and `use.effect`.

`deactivate` is handled via a new `registerFooBar` on the `WidgetSideEffectApi` (so you will need to do a `use.register()` to add a deactivate listener).

There are other lifecycle methods, but I haven't bothered to add support for them yet since I haven't had a need to (and they seem pretty unlikely to be needed when you're building an app with ReArch). If you happen to need one of these other methods, please raise an issue with a use case for it and I may add it in.

> I will start tonight to refactor my project Flutter side to use rearch but seems maybe a problem to not be able to have StatefulRearchWidget.

I don't think I want to add `StatefulRearchWidget` since it could encourage misusing side effects. While it can aid in migration, for now I'd just recommend the migration of a widget in its entirety every time.
